src/odemis/driver/orsay.py

```diff
@@ -2656,22 +2687,255 @@ def __init__(self, name, role, parent, **kwargs):
+        if aperture_number:
+            return int(aperture_number)
+        else:
+            raise LookupError(f"No aperture number preset found, None type was returned.")
+
+    def _getCondenserVoltageFromPreset(self, preset):
+        """
+        Get the condenser voltage from a given preset.
+
+        :param preset (xml.etree.ElementTree.Element): Full preset XML ElementTree including sub elements with all sub
+        components
+        :return (str): Value found for the setting
+        :raises LookupError if no condenser voltage is found for a preset
+        """
+        condenser_voltage = self.getPresetSetting(preset, 'HVPSFloatingIon', "CondensorVoltage", tag="Target")
+        if condenser_voltage:
+            return float(condenser_voltage)
+        else:
+            raise LookupError(f"No condenser voltage preset found, None type was returned.")
+
+    def _getScale(self):
+        image_format = self.parent._fib_beam.imageFormat.value
+        scale = (self.shape[0] / image_format[0],
+                self.shape[1] / image_format[1])
+        return scale
+
+    def _setScale(self, scale):
+        self.parent._fib_beam.imageFormat.value = (self.shape[0] / scale[0],
+                                                   self.shape[1] / scale[1])
+        return scale
+
+class Detector(model.Detector):
+    """
+    Represents the sensor for acquiring the image data.
+    """
+    SHAPE = (16384,)
+
+    def __init__(self, name, role, parent, **kwargs):
+        super(Detector, self).__init__(name, role, parent=parent, **kwargs)
+
+        if not hasattr(self.parent, "_fib_beam"):
+            raise ValueError(f"To create a Orsay Detecotr component the parent should also have the FIBBeam child.")
+        self._fib_beam = self.parent._fib_beam  # reference to the FIBBeam object
+
+        if not hasattr(self.parent, "_scanner"):
+            raise ValueError(f"To create a Orsay Detector component the parent should also have the scanner child.")
+        self._scanner = self.parent._scanner  # reference to the scanner object
+
+        self.contrast = self._fib_beam.contrast
+        self.brightness = self._fib_beam.brightness
+
+        self.data = Dataflow(self, self._fib_beam)
+
+    def receiveLatestImage(self):
+        """
+        Acquires the latest image from the Orsay server and converts it to a data array with type uint 16
+        :return (DataArray): Latest image acquired
+        """
+        metadata = self._metadata.copy()
+        metadata.update(self._scanner._metadata)
+        # TODO check if the estimate for MD_ACQ_DATE can be improved
+        scanning_time = self._scanner.dwellTime.value * self._scanner.resolution.value[0] * self._scanner.resolution.value[1]
+        metadata[model.MD_ACQ_DATE] = time.time() - scanning_time
+        image = self.parent.datamodel.Miss.AcquireFullImageScanOne()
+        uint16image = numpy.frombuffer(image, dtype=numpy.uint16).reshape(self._fib_beam.imageFormat.value)
```

Comment:
The reshape is tricky, because `imageFormat` might have changed since the image was acquired. In this case, it will fail, and you'll lose all the data. Is there any way to obtain the image shape? Admittedly, it's either 512x512 or 1024x1024, so it should be pretty easy to guess based on the size of the buffer!
This could be a reason why sometimes you've noticed that images are not received when changing settings.

Reply:
E